pkg/scheduler/ugm/manager.go

```diff
@@ -0,0 +1,176 @@
+
+	"github.com/apache/yunikorn-core/pkg/common/resources"
+	"github.com/apache/yunikorn-core/pkg/common/security"
+	"github.com/apache/yunikorn-core/pkg/log"
+)
+
+type Manager struct {
+	users  map[string]*UserTracker
+	groups map[string]*GroupTracker
+	sync.RWMutex
+}
+
+func NewManager() *Manager {
+	manager := &Manager{
+		users:  make(map[string]*UserTracker),
+		groups: make(map[string]*GroupTracker),
+	}
+	return manager
+}
+func (m *Manager) IncreaseTrackedResource(queuePath string, applicationID string, usage *resources.Resource, user *security.UserGroup) error {
+	if queuePath == "" || applicationID == "" || usage == nil || user == nil {
+		return fmt.Errorf("mandatory parameters are missing. queuepath: %s, application id: %s, resource usage: %s, user: %s",
+			queuePath, applicationID, usage.String(), user.User)
+	}
+	m.Lock()
+	defer m.Unlock()
+	var userTracker *UserTracker
+	if m.users[user.User] == nil {
+		userTracker = NewUserTracker(user)
+		m.users[user.User] = userTracker
+	} else {
+		userTracker = m.users[user.User]
+	}
+	err := userTracker.increaseTrackedResource(queuePath, applicationID, usage)
+	if err != nil {
+		return err
+	}
+	m.checkAppGroup(applicationID, user)
+	groupTracker := m.groups[m.getGroup(user)]
+	if groupTracker != nil {
+		err = groupTracker.increaseTrackedResource(queuePath, applicationID, usage)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (m *Manager) DecreaseTrackedResource(queuePath string, applicationID string, usage *resources.Resource, user *security.UserGroup, removeApp bool) error {
+	if queuePath == "" || applicationID == "" || usage == nil || user == nil {
+		return fmt.Errorf("mandatory parameters are missing. queuepath: %s, application id: %s, resource usage: %s, user: %s",
+			queuePath, applicationID, usage.String(), user.User)
+	}
+	m.Lock()
+	defer m.Unlock()
+	userTracker := m.users[user.User]
+	if userTracker != nil {
+		err := userTracker.decreaseTrackedResource(queuePath, applicationID, usage, removeApp)
+		if err != nil {
+			return err
+		}
+		if removeApp {
+			if m.isUserRemovable(userTracker) {
+				delete(m.users, user.User)
+			}
+		}
+	} else {
+		log.Logger().Info("user tracker must be available in users map",
+			zap.String("user", user.User))
+		return fmt.Errorf("user tracker for %s is missing in users map", user.User)
+	}
+
+	groupTracker := m.groups[m.getGroup(user)]
+	if groupTracker != nil {
+		err := groupTracker.decreaseTrackedResource(queuePath, applicationID, usage, removeApp)
+		if err != nil {
+			return err
+		}
+		if removeApp {
+			if m.isGroupRemovable(groupTracker) {
+				delete(m.groups, m.getGroup(user))
+			}
+		}
+	} else {
+		log.Logger().Info("group tracker must be available in groups map",
+			zap.String("group", m.getGroup(user)))
+		return fmt.Errorf("group tracker for %s is missing in groups map", m.getGroup(user))
+	}
+	return nil
+}
+
+func (m *Manager) checkAppGroup(applicationID string, user *security.UserGroup) {
```

Comment:
To me, it looks like this method makes sure that each group in the manager will have a `GroupTracker`. How about calling it `ensureGroupTrackerForApp()` or something similar?